File: source/com/barrybecker4/puzzle/tantrix/model/TilePlacement.java
// Copyright by Barry G. Becker, 2012. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.tantrix.model;

import com.barrybecker4.common.geometry.Location;

import java.util.HashMap;
import java.util.Map;

import static com.barrybecker4.puzzle.tantrix.model.HexTile.NUM_SIDES;

/**
 * Represents the positioning of a tantrix tile on the tantrix.
 * Immutable.
 * @author Barry Becker
 */
public class TilePlacement {

    private Location location;
    private Rotation rotation;
    private HexTile tile;

    /**
     * Constructor.
     */
    public TilePlacement(HexTile tile, Location location, Rotation rotation) {
        assert location != null;
        assert rotation != null;

        assert tile != null;
        this.tile = tile;
        this.location = location;
        this.rotation = rotation;
    }

    /**
     * Constructor for empty location.
     */
    public TilePlacement(Location location) {
        this(null, location, Rotation.ANGLE_0);
    }

    /**
     * Constructor.
     */
    public HexTile getTile() {
        return tile;
    }

    public PathColor getPathColor(int i) {
        int index = (i - rotation.ordinal()) % NUM_SIDES;
        index = (index < 0) ? index + NUM_SIDES : index;
        return tile.getEdgeColor(index);
    }

    public Location getLocation() {
        return location;
    }

    /** @return the amount that the tile is rotated. */
    public Rotation getRotation() {
        return rotation;
    }

    /**
     * Turn the new tile based on the old tile, but rotated counter-clockwise once.
     * @return new immutable TilePlacement instance.
     */
    public TilePlacement rotate() {
        Rotation newRotation = Rotation.values()[(rotation.ordinal() + 1) % NUM_SIDES];
        return new TilePlacement(tile, location, newRotation);
    }

    public Map<Integer, Location> getOutgoingPathLocations(PathColor primaryColor) {
        Map<Integer, Location> outgoingPathLocations = new HashMap<Integer, Location>();
        for (int i=0; i < NUM_SIDES; i++)  {
            if (primaryColor == getPathColor(i)) {
                outgoingPathLocations.put(i, HexUtil.getNeighborLocation(getLocation(), i));
            }
        }

        assert outgoingPathLocations.size() == 2: "There must always be two paths. Instead had "
                + outgoingPathLocations + " for "+this+" pcolor="+ primaryColor;
        return outgoingPathLocations;
    }


    public String toString() {
        return  "["  + tile +" at " + location + " " + rotation + "]";
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        TilePlacement that = (TilePlacement) o;
        if (location != null ? !location.equals(that.location) : that.location != null) return false;
        if (rotation != that.rotation) return false;
        if (tile != null ? !tile.equals(that.tile) : that.tile != null) return false;

        return true;
    }

    @Override
    public int hashCode() {
        int result = location != null ? location.hashCode() : 0;
        result = 31 * result + (rotation != null ? rotation.hashCode() : 0);
        result = 31 * result + (tile != null ? tile.hashCode() : 0);
        return result;
    }
}

File: scala-source/com/barrybecker4/puzzle/tantrix/model/TilePlacement.scala
// Copyright by Barry G. Becker, 2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.tantrix.model

import com.barrybecker4.common.geometry.Location
import com.barrybecker4.puzzle.tantrix.model.HexTile.NUM_SIDES
import com.barrybecker4.puzzle.tantrix.model.PathColor.PathColor
import com.barrybecker4.puzzle.tantrix.model.RotationEnum.Rotation

import scala.collection.immutable.HashMap


/**
  * Represents the positioning of a tantrix tile on the tantrix. Immutable.
  *
  * @author Barry Becker
  */
case class TilePlacement(tile: HexTile, location: Location, rotation: Rotation) {
  assert(location != null)
  assert(rotation != null)
  assert(tile != null)

  def getPathColor(i: Int): PathColor = {
    var index = (i - rotation.ordinal) % NUM_SIDES
    index = if (index < 0) index + NUM_SIDES
    else index
    tile.edgeColors(index)
  }

  /** @return new tile placement that is the old tile placement rotated counter-clockwise once. */
  def rotate(): TilePlacement = TilePlacement(tile, location, rotation.rotateBy(1))

  /**
    * @param primaryColor color of the path to get locations for
    * @return map from outgoing path index to corresponding location for paths of the specified color
    */
  def getOutgoingPathLocations(primaryColor: PathColor): Map[Int, Location] = {
    var outgoingPathLocations = new HashMap[Int, Location]
    for (i <- 0 until NUM_SIDES) {
      if (primaryColor == getPathColor(i))
        outgoingPathLocations += (i -> HexUtil.getNeighborLocation(location, i))
    }
    assert(outgoingPathLocations.size == 2,
      "Must always be two paths. Instead had " + outgoingPathLocations + " for " + this + " pcolor=" + primaryColor)
    outgoingPathLocations
  }

  override def toString: String = "[" + tile + " at " + location + " " + rotation + "]"
}